src/neo/SmartContract/Native/Oracle/OracleContract.cs

```diff
@@ -0,0 +1,178 @@
+            OracleResponse response = tx.Attributes.OfType<OracleResponse>().FirstOrDefault();
+            if (response is null) return tx.Hash;
+            OracleRequest request = GetRequest(engine.Snapshot, response.Id);
+            return request.OriginalTxid;
+        }
+
+        public OracleRequest GetRequest(StoreView snapshot, ulong id)
+        {
+            return snapshot.Storages.TryGet(CreateStorageKey(Prefix_Request).Add(id))?.GetInteroperable<OracleRequest>();
+        }
+
+        public IEnumerable<OracleRequest> GetRequests(StoreView snapshot)
+        {
+            return snapshot.Storages.Find(new byte[] { Prefix_Request }).Select(p => p.Value.GetInteroperable<OracleRequest>());
+        }
+
+        public IEnumerable<OracleRequest> GetRequestsByUrl(StoreView snapshot, string url)
+        {
+            IdList list = snapshot.Storages.TryGet(CreateStorageKey(Prefix_IdList).Add(GetUrlHash(url)))?.GetInteroperable<IdList>();
+            if (list is null) yield break;
+            foreach (ulong id in list)
+                yield return snapshot.Storages[CreateStorageKey(Prefix_Request).Add(id)].GetInteroperable<OracleRequest>();
+        }
+
+        private static byte[] GetUrlHash(string url)
+        {
+            return Crypto.Hash160(Utility.StrictUTF8.GetBytes(url));
+        }
+
+        internal override void Initialize(ApplicationEngine engine)
+        {
+            engine.Snapshot.Storages.Add(CreateStorageKey(Prefix_NodeList), new StorageItem(new NodeList()));
+            engine.Snapshot.Storages.Add(CreateStorageKey(Prefix_RequestId), new StorageItem(BitConverter.GetBytes(0ul)));
+        }
+
+        protected override void PostPersist(ApplicationEngine engine)
+        {
+            base.PostPersist(engine);
+            foreach (Transaction tx in engine.Snapshot.PersistingBlock.Transactions)
+            {
+                //Filter the response transactions
+                OracleResponse response = tx.Attributes.OfType<OracleResponse>().FirstOrDefault();
+                if (response is null) continue;
+
+                //Remove the request from storage
+                StorageKey key = CreateStorageKey(Prefix_Request).Add(response.Id);
+                OracleRequest request = engine.Snapshot.Storages[key].GetInteroperable<OracleRequest>();
+                engine.Snapshot.Storages.Delete(key);
+
+                //Remove the id from IdList
+                key = CreateStorageKey(Prefix_IdList).Add(GetUrlHash(request.Url));
+                IdList list = engine.Snapshot.Storages.GetAndChange(key).GetInteroperable<IdList>();
+                if (!list.Remove(response.Id)) throw new InvalidOperationException();
+                if (list.Count == 0) engine.Snapshot.Storages.Delete(key);
+
+                //Mint GAS for oracle nodes
+                ECPoint[] nodes = GetOracleNodes(engine.Snapshot);
+                int index = (int)(response.Id % (ulong)nodes.Length);
```

Comment:
if nodes are empty, it could divide by 0

Reply:
Fixed.